File: 22FALL/130.surrounded-regions.py
#
# @lc app=leetcode id=130 lang=python3
#
# [130] Surrounded Regions
#

# @lc code=start
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        m, n =  len(board), len(board[0])
        directions = [0, 1, 0, -1, 0]

        def dfs(i, j):
            if i < 0 or i > m-1 or j < 0 or j > n-1 or board[i][j] != 'O':
                return
            board[i][j] = 'Y'
            for k in range(4):
                dx, dy = directions[k], directions[k+1]
                x, y = i + dx, j + dy
                dfs(x, y)
        
        for i in range(m):
            dfs(i, 0)
            dfs(i, n-1)
        for j in range(n-1):
            dfs(0, j)
            dfs(m-1, j)
        for i in range(m):
            for j in range(n):
                if board[i][j] == 'Y':
                    board[i][j] = 'O'
                elif board[i][j] == 'O':
                    board[i][j] = 'X'
        return 

# @lc code=end

File: VSCode_CPP/130.surrounded-regions.cpp
/*
 * @lc app=leetcode id=130 lang=cpp
 *
 * [130] Surrounded Regions
 */

// @lc code=start
class Solution {
public:
    void solve(vector<vector<char>>& board) {
        m = board.size();
        n = board[0].size();
        for (int i=0; i<m; i++) {
            if (board[i][0] == 'O') dfs(board, i, 0);
            if (board[i][n-1] == 'O') dfs(board, i, n-1);
        }
        for (int j=0; j<n; j++) {
            if (board[0][j] == 'O') dfs(board, 0, j);
            if (board[m-1][j] == 'O') dfs(board, m-1, j);
        }
        for (int i=0; i<m; i++){
            for (int j=0; j<n; j++) {
                if (board[i][j] == 'O') board[i][j] = 'X';
                if (board[i][j] == 'T') board[i][j] = 'O';
            }
        }
    }
private:
    int m, n;
    void dfs(vector<vector<char>>& board, int i, int j) {
        if (i<0 || i>=m || j<0 || j>=n || board[i][j] != 'O') return;
        board[i][j] = 'T';
        dfs(board, i+1, j);
        dfs(board, i-1, j);
        dfs(board, i, j+1);
        dfs(board, i, j-1);
    }
};
// @lc code=end

